### PR TITLE
Cuckoo Hashtable & Cuckoo Hashtable tests

### PR DESCRIPTION
**Pull request by Daniel-ND**

### CODE REVIEW

**Review by kurapov-peter — CHANGES_REQUESTED** [review]
Very nice overall, need a couple cosmetic changes

File: common/dpcpp/cuckoo_hashtable.hpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                return {_vals[pos1], true};
+            if (_keys[pos2] == key)
+                return {_vals[pos2], true};
+            return {{}, false};
+        }
+
+        const bool has(Key key) {
+             if (_keys[_hasher1(key)] == key || _keys[_hasher2(key)] == key)
+                return true;
+            return false;
```
[the remark this answers is not in the document]
Comment (Daniel-ND) [reply common/dpcpp/cuckoo_hashtable.hpp]: fixed

File: common/dpcpp/cuckoo_hashtable.hpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        bool insert(Key key, Val value, size_t cnt) {
```
[the remark this answers is not in the document]
Comment (Daniel-ND) [reply common/dpcpp/cuckoo_hashtable.hpp]: By mistake left from when this method was recursive. Removed.

File: common/dpcpp/cuckoo_hashtable.hpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            auto pos1 = _hasher1(key);
+            auto pos2 = _hasher2(key);
+            if (_keys[pos1] == key)
+                return {_vals[pos1], true};
+            if (_keys[pos2] == key)
+                return {_vals[pos2], true};
+            return {{}, false};
+        }
+
+        const bool has(Key key) {
```
[the remark this answers is not in the document]
Comment (Daniel-ND) [reply common/dpcpp/cuckoo_hashtable.hpp]: Made this method const, by this logic, the method at() also. They both should't modify hashtable.

File: common/dpcpp/cuckoo_hashtable.hpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+             if (_keys[_hasher1(key)] == key || _keys[_hasher2(key)] == key)
+                return true;
+            return false;
+        }
+
+        bool insert(Key key, Val value, size_t cnt) {
+            // TODO: change loop detection
+            while (cnt < _size) {
+
+                uint32_t mask, present, major_idx;
```
[the remark this answers is not in the document]
Comment (Daniel-ND) [reply common/dpcpp/cuckoo_hashtable.hpp]: Moved these variables to lock & unlock methods. Think that's better.

File: common/dpcpp/cuckoo_hashtable.hpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+template <class Key, class Val, class Hasher1, class Hasher2> class CuckooHashtable{
+    private:
+        sycl::global_ptr<Key> _keys;
+        sycl::global_ptr<Val> _vals;
+
+        const size_t _size;
+        Hasher1 _hasher1;
+        Hasher2 _hasher2;
+        const Key _EMPTY_KEY;
```
[the remark this answers is not in the document]
Comment (Daniel-ND) [reply common/dpcpp/cuckoo_hashtable.hpp]: fixed

File: tests/cuckoo_hashtable_tests.cpp [hunk 6]
```diff
@@ -2,7 +2,7 @@
 #include <gtest/gtest.h>
 #include <vector>
 
-#define EMPTY_KEY 2147483647
+#define EMPTY_KEY std::numeric_limits<uint32_t>::max()
```
Comment (kurapov-peter) [inline tests/cuckoo_hashtable_tests.cpp:5]: Unused?

**Review by kurapov-peter — APPROVED** [review]
LGTM

**Pull request merged**